packages/field-base/src/field-mixin.js

```diff
@@ -103,49 +99,106 @@ export const FieldMixin = (superclass) =>
     ready() {
       super.ready();
 
-      if (this._errorNode) {
-        this._errorNode.id = this._errorId;
+      const error = this._errorNode;
+      if (error) {
+        error.id = this._errorId;
 
-        this._applyCustomError();
+        this.__applyCustomError();
 
         this._updateErrorMessage(this.invalid, this.errorMessage);
       }
 
-      if (this._helperNode) {
-        this._currentHelper = this._helperNode;
-        this._helperNode.id = this._helperId;
-
-        this._applyCustomHelper();
-
-        this.__helperSlot = this.shadowRoot.querySelector('[name="helper"]');
-        this.__helperSlot.addEventListener('slotchange', this.__onHelperSlotChange.bind(this));
-        this.__helperIdObserver = new MutationObserver((mutationRecord) => {
-          mutationRecord.forEach((mutation) => {
-            // only handle helper nodes
-            if (
-              mutation.type === 'attributes' &&
-              mutation.attributeName === 'id' &&
-              mutation.target === this._currentHelper &&
-              mutation.target.id !== this.__savedHelperId
-            ) {
-              this.__updateHelperId(mutation.target);
-            }
+      const helper = this._helperNode;
+      if (helper) {
+        this.__applyCustomHelper(helper);
+      }
+
+      this.__helperSlot = this.shadowRoot.querySelector('[name="helper"]');
+
+      this._observer = new FlattenedNodesObserver(this.__helperSlot, (info) => {
+        const helper = this._currentHelper;
+
+        const newHelper = info.addedNodes.find((node) => node !== helper);
+        const oldHelper = info.removedNodes.find((node) => node === helper);
+
+        if (newHelper) {
+          // Custom helper is added, remove the previous one.
+          if (helper && helper.isConnected) {
+            this.removeChild(helper);
+          }
+
+          this.__applyCustomHelper(newHelper);
+
+          this.__helperIdObserver = new MutationObserver((mutations) => {
+            mutations.forEach((mutation) => {
+              // only handle helper nodes
+              if (
+                mutation.type === 'attributes' &&
+                mutation.attributeName === 'id' &&
+                mutation.target === this._currentHelper &&
+                mutation.target.id !== this.__savedHelperId
+              ) {
+                this.__updateHelperId(mutation.target);
+              }
+            });
           });
-        });
 
-        this.__helperIdObserver.observe(this, { attributes: true, subtree: true });
-      }
+          this.__helperIdObserver.observe(newHelper, { attributes: true });
+        } else if (oldHelper) {
+          this.__helperIdObserver.disconnect();
+
+          if (this.helperText) {
+            // Custom helper is removed, restore the default one.
+            const helper = this.__attachDefaultHelper();
```

Comment:
Oh, I've just happened to notice that the `has-helper` attribute isn't covered here.

In the case of restoring the default helper, we need to take care of the `has-helper` attribute. Also, we might want to set the helper text only if it is not empty like it is done in `_helperTextChanged`.

Reply:
The component seems to produce an error if the `<label slot="helper">` gets removed dynamically.

![Screenshot 2021-10-04 at 11 01 52](https://user-images.githubusercontent.com/1222264/135815018-932d60f9-0457-466c-8a87-d95ba3e673f0.png)

Reply:
Updated the code to reuse logic from the observer. @tomivirkki please check if this still happens.

Reply:
The error still occurs

https://user-images.githubusercontent.com/1222264/135832183-18f1d825-7b05-4cb4-9af9-db3f85fbaa29.mp4

Reply:
@tomivirkki thanks, now it is fixed.